In [1]:
%load_ext autoreload
%autoreload 2

In [36]:
# Imports
import numpy as np
import matplotlib.pyplot as plt
import keras 
from keras.layers import Input, Dense
from sklearn.model_selection import train_test_split
from src.data_utils import load_fizzbuzz_csv, _cast_list_to_type, fizzbuzz_to_formatted_array, evaluate_model, evaluate_binary_model

# Load and Preprocess Data
Large quantities of data are required to train any neural network. While manually creating a dataset like this is a useful exercise, I have gone ahead and hand-labeled the first 100,000 examples. These are stored in a file called fizzbuzzdata.csv in the data directory. <br>
So, let's load up that data, preview it, and get ready to process it.

In [9]:
# Load data
x_data, y_data = load_fizzbuzz_csv()

# Preview data
for i in range(15):
    print("x: %s | y: %s"%(x_data[i], y_data[i]))

x: 1 | y:  1
x: 2 | y:  2
x: 3 | y:  Fizz
x: 4 | y:  4
x: 5 | y:  Buzz
x: 6 | y:  Fizz
x: 7 | y:  7
x: 8 | y:  8
x: 9 | y:  Fizz
x: 10 | y:  Buzz
x: 11 | y:  11
x: 12 | y:  Fizz
x: 13 | y:  13
x: 14 | y:  14
x: 15 | y:  FizzBuzz


Our data looks good, so let's start getting it ready for our neural network. For our first example, we'll be using a multilayer perceptron with in input dimension of 1 and an output dimension of 2. We'll set it up so that the first dimension of the output vector represents the probability that the input should result in a "Fizz", and the second dimension represents the probability that the input should result in a "Buzz".

In [10]:
# Process Data
# X DATA
# Make sure that x data is integers
x_data = _cast_list_to_type(x_data, int)
x_data = np.array(x_data)

# Y DATA (onehot)
y_data = fizzbuzz_to_formatted_array(y_data)

# Check data shapes
print("X data shape:", x_data.shape)
print("Y data shape:", y_data.shape)

X data shape: (100000,)
Y data shape: (100000, 2)


Now we'll split our data into training/testing splits so that we can properly judge the quality of our model.

In [11]:
x_train, x_test, y_train, y_test = train_test_split(x_data, y_data, test_size=0.33)
# Check data shapes
print("X train shape:", x_train.shape)
print("X test shape:", x_test.shape)
print("Y train shape:", y_train.shape)
print("Y test shape:", y_test.shape)

X train shape: (67000,)
X test shape: (33000,)
Y train shape: (67000, 2)
Y test shape: (33000, 2)


# Define, Train, and Evaluate model
Now, we'll define a simple neural network that should be capable of solving this problem. If we don't see the accuracy we want, we'll simply add more layers later.

In [12]:
# Define Model
input_layer = Input(1)
x = Dense(10, activation='relu')(input_layer)
x = Dense(100, activation='relu')(x)
x = Dense(10, activation='relu')(x)
output = Dense(2, activation='sigmoid')(x)

fizzbuzzmodel = keras.Model(inputs=input_layer, outputs=output)
fizzbuzzmodel.summary()

Model: "functional_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
input_1 (InputLayer)         [(None, 1)]               0         
_________________________________________________________________
dense (Dense)                (None, 10)                20        
_________________________________________________________________
dense_1 (Dense)              (None, 100)               1100      
_________________________________________________________________
dense_2 (Dense)              (None, 10)                1010      
_________________________________________________________________
dense_3 (Dense)              (None, 2)                 22        
Total params: 2,152
Trainable params: 2,152
Non-trainable params: 0
_________________________________________________________________


Now it's finally time to train our neural network!

In [14]:
# Train Model
optimizer = keras.optimizers.Adam(learning_rate=0.0001)
fizzbuzzmodel.compile(optimizer=optimizer, loss='mse', metrics=['accuracy', keras.metrics.Precision(), keras.metrics.Recall()])
history = fizzbuzzmodel.fit(x_train, y_train, epochs=100, batch_size=128)

Epoch 1/100
524/524 [==============================] - 1s 1ms/step - loss: 0.1911 - accuracy: 0.8664 - precision_1: 0.0000e+00 - recall_1: 0.0000e+00
Epoch 2/100
524/524 [==============================] - 1s 1ms/step - loss: 0.1911 - accuracy: 0.8664 - precision_1: 0.0000e+00 - recall_1: 0.0000e+00
Epoch 3/100
524/524 [==============================] - 1s 1ms/step - loss: 0.1911 - accuracy: 0.8664 - precision_1: 0.0000e+00 - recall_1: 0.0000e+00
Epoch 4/100
524/524 [==============================] - 1s 981us/step - loss: 0.1911 - accuracy: 0.8664 - precision_1: 0.0000e+00 - recall_1: 0.0000e+00
Epoch 5/100
524/524 [==============================] - 1s 1ms/step - loss: 0.1911 - accuracy: 0.8664 - precision_1: 0.0000e+00 - recall_1: 0.0000e+00
Epoch 6/100
524/524 [==============================] - 1s 1ms/step - loss: 0.1911 - accuracy: 0.8664 - precision_1: 0.0000e+00 - recall_1: 0.0000e+00
Epoch 7/100
524/524 [==============================] - 1s 1ms/step - loss: 0.1911 - accuracy: 0.86

In [26]:
# Evaluate Model
evaluate_model(fizzbuzzmodel, 1000)

0.533

Looks like we're getting an accuracy of 53%. That's not bad, we're already doing better than a coin flip! But we can do better-- it's time to bring out the big guns. <br>
Let's load up a bigger neural network and see how it does!

In [27]:
# Define Model
input_layer = Input(1)
x = Dense(10, activation='relu')(input_layer)
x = Dense(100, activation='relu')(x)
x = Dense(100, activation='relu')(x)
x = Dense(100, activation='relu')(x)
x = Dense(100, activation='relu')(x)
x = Dense(100, activation='relu')(x)
x = Dense(10, activation='relu')(x)
output = Dense(2, activation='sigmoid')(x)

fizzbuzzmodel = keras.Model(inputs=input_layer, outputs=output)
optimizer = keras.optimizers.Adam(learning_rate=0.0001)
fizzbuzzmodel.compile(optimizer=optimizer, loss='mse', metrics=['accuracy', keras.metrics.Precision(), keras.metrics.Recall()])
history = fizzbuzzmodel.fit(x_train, y_train, epochs=100, batch_size=128)

Epoch 1/100
524/524 [==============================] - 1s 2ms/step - loss: 0.2445 - accuracy: 0.8267 - precision_2: 0.3428 - recall_2: 0.0310
Epoch 2/100
524/524 [==============================] - 1s 3ms/step - loss: 0.2159 - accuracy: 0.8664 - precision_2: 0.0000e+00 - recall_2: 0.0000e+00
Epoch 3/100
524/524 [==============================] - 1s 3ms/step - loss: 0.2145 - accuracy: 0.8663 - precision_2: 0.0000e+00 - recall_2: 0.0000e+00
Epoch 4/100
524/524 [==============================] - 1s 3ms/step - loss: 0.2361 - accuracy: 0.2677 - precision_2: 0.2148 - recall_2: 0.0015
Epoch 5/100
524/524 [==============================] - 1s 2ms/step - loss: 0.2330 - accuracy: 0.1336 - precision_2: 0.0000e+00 - recall_2: 0.0000e+00
Epoch 6/100
524/524 [==============================] - 1s 3ms/step - loss: 0.2276 - accuracy: 0.1336 - precision_2: 0.0000e+00 - recall_2: 0.0000e+00
Epoch 7/100
524/524 [==============================] - 1s 2ms/step - loss: 0.2232 - accuracy: 0.1336 - precision_2: 

In [28]:
# Evaluate Model
evaluate_model(fizzbuzzmodel, 1000)

0.533

Hmm, looks like we didn't see much improvement. One more time, let's try deeper!

In [29]:
# Define Model
input_layer = Input(1)
x = Dense(10, activation='relu')(input_layer)
x = Dense(100, activation='relu')(x)
x = Dense(100, activation='relu')(x)
x = Dense(1000, activation='relu')(x)
x = Dense(1000, activation='relu')(x)
x = Dense(1000, activation='relu')(x)
x = Dense(100, activation='relu')(x)
x = Dense(100, activation='relu')(x)
x = Dense(10, activation='relu')(x)
output = Dense(2, activation='sigmoid')(x)

fizzbuzzmodel = keras.Model(inputs=input_layer, outputs=output)
optimizer = keras.optimizers.Adam(learning_rate=0.0001)
fizzbuzzmodel.compile(optimizer=optimizer, loss='mse', metrics=['accuracy', keras.metrics.Precision(), keras.metrics.Recall()])
history = fizzbuzzmodel.fit(x_train, y_train, epochs=100, batch_size=128)

Epoch 1/100
524/524 [==============================] - 9s 17ms/step - loss: 0.3122 - accuracy: 0.7486 - precision_3: 0.2945 - recall_3: 0.1245
Epoch 2/100
524/524 [==============================] - 9s 18ms/step - loss: 0.2746 - accuracy: 0.8285 - precision_3: 0.3264 - recall_3: 0.0286
Epoch 3/100
524/524 [==============================] - 10s 18ms/step - loss: 0.2847 - accuracy: 0.8535 - precision_3: 0.3317 - recall_3: 0.0677
Epoch 4/100
524/524 [==============================] - 9s 17ms/step - loss: 0.2693 - accuracy: 0.8664 - precision_3: 0.3356 - recall_3: 0.0108
Epoch 5/100
524/524 [==============================] - 9s 18ms/step - loss: 0.3283 - accuracy: 0.8664 - precision_3: 0.3345 - recall_3: 0.2343
Epoch 6/100
524/524 [==============================] - 10s 19ms/step - loss: 0.3254 - accuracy: 0.8411 - precision_3: 0.3307 - recall_3: 0.2156
Epoch 7/100
524/524 [==============================] - 10s 18ms/step - loss: 0.2717 - accuracy: 0.8205 - precision_3: 0.3341 - recall_3: 0.0

KeyboardInterrupt: 

In [30]:
# Evaluate Model
evaluate_model(fizzbuzzmodel, 1000)

0.533

Looks like we might need a new approach.<br>
Before we abandon the idea of using AI at all, let's think outside the box. Maybe there's a way for us to increase the dimensionality of our data, giving the network more to work with? I struggle to motivate these next steps, but stick with me and you'll see where we're going.

## Binary Representation

In [31]:
# We'll use this function to reformat our input data from a one-dimensional integer to a 30-dimensional vector.
input_dim = 30

def x_data_reformatter(n, input_dim):
    bin_n = bin(n)
    str_n = bin_n[2:]
    if len(str_n) < input_dim:
        str_n = '0'*(input_dim-len(str_n)) + str_n 
    return [int(char) for char in str_n]

reformatted_x_data = []
for i in x_data:
    reformatted_x_data.append(x_data_reformatter(i, input_dim))
reformatted_x_data = np.array(reformatted_x_data)

In [32]:
# Train new model
input_layer = Input(input_dim)
x = Dense(10, activation='relu')(input_layer)
x = Dense(100, activation='relu')(x)
x = Dense(10, activation='relu')(x)
output = Dense(2, activation='sigmoid')(x)
fizzbuzzmodel = keras.Model(inputs=input_layer, outputs=output)
optimizer = keras.optimizers.Adam(learning_rate=0.0001)
fizzbuzzmodel.compile(optimizer=optimizer, loss='mse', metrics=['accuracy', keras.metrics.Precision(), keras.metrics.Recall()])
fizzbuzzmodel.fit(reformatted_x_data, y_data, epochs=1000, batch_size=256)

Epoch 1/1000
391/391 [==============================] - 1s 1ms/step - loss: 0.2180 - accuracy: 0.7066 - precision_4: 0.2127 - recall_4: 0.0365
Epoch 2/1000
391/391 [==============================] - 1s 1ms/step - loss: 0.1932 - accuracy: 0.8667 - precision_4: 0.0000e+00 - recall_4: 0.0000e+00
Epoch 3/1000
391/391 [==============================] - 1s 1ms/step - loss: 0.1926 - accuracy: 0.8667 - precision_4: 0.0000e+00 - recall_4: 0.0000e+00
Epoch 4/1000
391/391 [==============================] - 1s 1ms/step - loss: 0.1924 - accuracy: 0.8667 - precision_4: 0.0000e+00 - recall_4: 0.0000e+00
Epoch 5/1000
391/391 [==============================] - 1s 1ms/step - loss: 0.1922 - accuracy: 0.8667 - precision_4: 0.0000e+00 - recall_4: 0.0000e+00
Epoch 6/1000
391/391 [==============================] - 0s 1ms/step - loss: 0.1920 - accuracy: 0.8667 - precision_4: 0.0000e+00 - recall_4: 0.0000e+00
Epoch 7/1000
391/391 [==============================] - 1s 1ms/step - loss: 0.1918 - accuracy: 0.8667 

In [35]:
evaluate_binary_model(fizzbuzzmodel, 10000)

0.964

Wow! 96% Accuracy!
With results like these, it's hard to argue against the efficacy of neural networks for everyday tasks.